File: lib/challengeUtils.ts
import config from 'config'
import { Op } from 'sequelize'
import colors from 'colors/safe'
import { type Server } from 'socket.io'
import sanitizeHtml from 'sanitize-html'
import { AllHtmlEntities as Entities } from 'html-entities'

import { type ChallengeKey, ChallengeModel } from '../models/challenge'
import { challenges, notifications } from '../data/datacache'
import { HintModel } from '../models/hint'
import * as accuracy from './accuracy'
import * as webhook from './webhook'
import * as antiCheat from './antiCheat'
import * as utils from './utils'
import logger from './logger'

const entities = new Entities()

const globalWithSocketIO = global as typeof globalThis & {
  io: Server
}

export const solveIf = function (challenge: any, criteria: () => any, isRestore: boolean = false, isCheating = false) {
  if (notSolved(challenge) && criteria()) {
    solve(challenge, isRestore, isCheating)
  }
}

export const solve = async function (challenge: ChallengeModel, isRestore = false, isCheating = false) {
  challenge.solved = true
  const solvedChallenge = await challenge.save()

  logger.info(`${isRestore ? colors.grey('Restored') : colors.green('Solved')} ${solvedChallenge.difficulty}-star ${colors.cyan(solvedChallenge.key)} (${solvedChallenge.name})`)
  sendNotification(solvedChallenge, isRestore)
  if (!isRestore) {
    const cheatScore = antiCheat.calculateCheatScore(challenge, isCheating)
    const hintsAvailable = await HintModel.count({ where: { ChallengeId: solvedChallenge.id } })
    const hintsUnlocked = await HintModel.count({ where: { ChallengeId: solvedChallenge.id, unlocked: true } })
    if (process.env.SOLUTIONS_WEBHOOK) {
      webhook.notify(solvedChallenge, cheatScore, hintsAvailable, hintsUnlocked).catch((error: unknown) => {
        logger.error('Webhook notification failed: ' + colors.red(utils.getErrorMessage(error)))
      })
    }
  }
}

export const sendNotification = function (challenge: ChallengeModel, isRestore: boolean) {
  if (notSolved(challenge)) {
    return
  }

  const flag = utils.ctfFlag(challenge.name)

  const challengeKey = challenge.key as ChallengeKey
  const fullChallenge = challenges[challengeKey]

  let hasCodingChallenge = false
  if (fullChallenge) {
    hasCodingChallenge = Boolean(fullChallenge.hasCodingChallenge) ?? false
  }

  const notification = {
    key: challenge.key,
    name: challenge.name,
    challenge: challenge.name + ' (' + entities.decode(sanitizeHtml(challenge.description, { allowedTags: [], allowedAttributes: {} })) + ')',
    flag,
    hidden: !config.get('challenges.showSolvedNotifications'),
    isRestore,
    codingChallenge: config.get('challenges.codingChallengesEnabled') !== 'never' && hasCodingChallenge
  }
  const wasPreviouslyShown = notifications.some(({ key }) => key === challenge.key)
  notifications.push(notification)

  if (globalWithSocketIO.io && (isRestore || !wasPreviouslyShown)) {
    globalWithSocketIO.io.emit('challenge solved', notification)
  }
}

export const sendCodingChallengeNotification = function (challenge: { key: string, codingChallengeStatus: 0 | 1 | 2 }) {
  if (challenge.codingChallengeStatus > 0) {
    const notification = {
      key: challenge.key,
      codingChallengeStatus: challenge.codingChallengeStatus
    }
    if (globalWithSocketIO.io) {
      globalWithSocketIO.io.emit('code challenge solved', notification)
    }
  }
}

export const notSolved = (challenge: ChallengeModel) => challenge && !challenge.solved

export const findChallengeByName = (challengeName: string) => {
  for (const challenge of Object.values(challenges)) {
    if (challenge.name === challengeName) {
      return challenge
    }
  }
  logger.warn('Missing challenge with name: ' + challengeName)
}

export const findChallengeById = (challengeId: number) => {
  for (const challenge of Object.values(challenges)) {
    if (challenge.id === challengeId) {
      return challenge
    }
  }
  logger.warn('Missing challenge with id: ' + challengeId)
}

export const solveFindIt = async function (key: ChallengeKey, isRestore: boolean = false) {
  const solvedChallenge = challenges[key]
  await ChallengeModel.update({ codingChallengeStatus: 1 }, { where: { key, codingChallengeStatus: { [Op.lt]: 2 } } })
  logger.info(`${isRestore ? colors.grey('Restored') : colors.green('Solved')} 'Find It' phase of coding challenge ${colors.cyan(solvedChallenge.key)} (${solvedChallenge.name})`)
  if (!isRestore) {
    accuracy.storeFindItVerdict(solvedChallenge.key, true)
    accuracy.calculateFindItAccuracy(solvedChallenge.key)
    await antiCheat.calculateFindItCheatScore(solvedChallenge)
    sendCodingChallengeNotification({ key, codingChallengeStatus: 1 })
  }
}

export const solveFixIt = async function (key: ChallengeKey, isRestore: boolean = false) {
  const solvedChallenge = challenges[key]
  await ChallengeModel.update({ codingChallengeStatus: 2 }, { where: { key } })
  logger.info(`${isRestore ? colors.grey('Restored') : colors.green('Solved')} 'Fix It' phase of coding challenge ${colors.cyan(solvedChallenge.key)} (${solvedChallenge.name})`)
  if (!isRestore) {
    accuracy.storeFixItVerdict(solvedChallenge.key, true)
    accuracy.calculateFixItAccuracy(solvedChallenge.key)
    await antiCheat.calculateFixItCheatScore(solvedChallenge)
    sendCodingChallengeNotification({ key, codingChallengeStatus: 2 })
  }
}

File: lib/challenge_utils.py
import logging
import config
import utils
import sanitize_html
import colors
import webhook
import accuracy
from models.challenge import ChallengeModel
from data.datacache import challenges, notifications
from anti_cheat import calculate_cheat_score, calculate_find_it_cheat_score, calculate_fix_it_cheat_score
from socketio import Server
from html import unescape

logger = logging.getLogger(__name__)

def solve_if(challenge, criteria, is_restore=False):
    if not_solved(challenge) and criteria():
        solve(challenge, is_restore)

def solve(challenge, is_restore=False):
    challenge.solved = True
    challenge.save().then(lambda solved_challenge: on_challenge_solved(solved_challenge, is_restore))

def on_challenge_solved(solved_challenge, is_restore):
    logger.info(f"{'Restored' if is_restore else 'Solved'} {solved_challenge.difficulty}-star {colors.cyan(solved_challenge.key)} ({solved_challenge.name})")
    send_notification(solved_challenge, is_restore)
    if not is_restore:
        cheat_score = calculate_cheat_score(solved_challenge)
        if process.env.SOLUTIONS_WEBHOOK:
            webhook.notify(solved_challenge, cheat_score).catch(lambda error: logger.error(f"Webhook notification failed: {colors.red(utils.get_error_message(error))}"))

def send_notification(challenge, is_restore):
    if not not_solved(challenge):
        flag = utils.ctf_flag(challenge.name)
        notification = {
            'key': challenge.key,
            'name': challenge.name,
            'challenge': f"{challenge.name} ({unescape(sanitize_html(challenge.description, allowed_tags=[], allowed_attributes={}))})",
            'flag': flag,
            'hidden': not config.get('challenges.showSolvedNotifications'),
            'is_restore': is_restore
        }
        was_previously_shown = any(n['key'] == challenge.key for n in notifications)
        notifications.append(notification)

        if global_with_socket_io.io and (is_restore or not was_previously_shown):
            global_with_socket_io.io.emit('challenge solved', notification)

def send_coding_challenge_notification(challenge):
    if challenge.coding_challenge_status > 0:
        notification = {
            'key': challenge.key,
            'coding_challenge_status': challenge.coding_challenge_status
        }
        if global_with_socket_io.io:
            global_with_socket_io.io.emit('code challenge solved', notification)

def not_solved(challenge):
    return challenge and not challenge.solved

def find_challenge_by_name(challenge_name):
    for c in challenges:
        if challenges[c].name == challenge_name:
            return challenges[c]
    logger.warn(f"Missing challenge with name: {challenge_name}")

def find_challenge_by_id(challenge_id):
    for c in challenges:
        if challenges[c].id == challenge_id:
            return challenges[c]
    logger.warn(f"Missing challenge with id: {challenge_id}")

def solve_find_it(key, is_restore):
    solved_challenge = challenges[key]
    ChallengeModel.update({'coding_challenge_status': 1}, {'where': {'key': key, 'coding_challenge_status': {'$lt': 2}}})
    logger.info(f"{'Restored' if is_restore else 'Solved'} 'Find It' phase of coding challenge {colors.cyan(solved_challenge.key)} ({solved_challenge.name})")
    if not is_restore:
        accuracy.store_find_it_verdict(solved_challenge.key, True)
        accuracy.calculate_find_it_accuracy(solved_challenge.key)
        calculate_find_it_cheat_score(solved_challenge)
        send_coding_challenge_notification({'key': key, 'coding_challenge_status': 1})

def solve_fix_it(key, is_restore):
    solved_challenge = challenges[key]
    ChallengeModel.update({'coding_challenge_status': 2}, {'where': {'key': key}})
    logger.info(f"{'Restored' if is_restore else 'Solved'} 'Fix It' phase of coding challenge {colors.cyan(solved_challenge.key)} ({solved_challenge.name})")
    if not is_restore:
        accuracy.store_fix_it_verdict(solved_challenge.key, True)
        accuracy.calculate_fix_it_accuracy(solved_challenge.key)
        calculate_fix_it_cheat_score(solved_challenge)
        send_coding_challenge_notification({'key': key, 'coding_challenge_status': 2})
